### PR TITLE
[opt] Flatten if(0) and if(1)

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue = #1372

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: tests/python/test_while.py [hunk 1]
```diff
@@ -38,4 +38,4 @@ def func():
         ret[None] = s
 
     func()
-    print(ret[None])
+    assert ret[None] == 55
```
Comment (yuanming-hu) [inline tests/python/test_while.py:41]: Thanks for fixing this ;-)

**Comment by k-ye** [comment]
At least on Metal, this broke https://github.com/taichi-dev/taichi/blob/b709fb63fb0d5b5ce3a34c042bb8d3f7148f6f0a/tests/python/test_tuple_assign.py#L9-L12

I think the `if 1` in that test is to explicitly make the for loop serial, but now it's been promoted into a `range-for` kernel.

* IR with this PR:

```
kernel {
  $0 = offloaded  
  body {
    <i32*x1> $1 = global tmp var (offset = 8 B)
    <i32 x1> $2 = const [0]
    <i32*x1> $3 : global store [$1 <- $2]
    <i32*x1> $4 = global tmp var (offset = 4 B)
    <i32 x1> $5 = const [1]
    <i32*x1> $6 : global store [$4 <- $5]
    <i32 x1> $7 = arg[0]
    <i32*x1> $8 = global tmp var (offset = 0 B)
    <i32*x1> $9 : global store [$8 <- $7]
  }
  $10 = offloaded range_for(0, tmp(offset=0B)) block_dim=adaptive  
  body {
    <i32*x1> $11 = global tmp var (offset = 4 B)
    <i32 x1> $12 = global load $11
    <i32*x1> $13 = global tmp var (offset = 8 B)
    <i32 x1> $14 = global load $13
    <i32 x1> $15 = add $14 $12
    <i32*x1> $16 : global store [$13 <- $12]
    <i32*x1> $17 : global store [$11 <- $15]
  }
  $18 = offloaded  
  body {
    <i32*x1> $19 = global tmp var (offset = 4 B)
    <i32 x1> $20 = global load $19
    <i32 x1> $21 : kernel return $20
  }
}
```

* IR before

```
kernel {
  $0 = offloaded  
  body {
    <i32 x1> $1 = const [0]
    <i32 x1> $2 = const [1]
    <i32 x1> $3 = alloca
    <i32 x1> $4 : local store [$3 <- $1]
    <i32 x1> $5 = alloca
    <i32 x1> $6 : local store [$5 <- $2]
    $7 : if $2 {
      <i32 x1> $8 = arg[0]
      $9 : for in range($1, $8) (vectorize 1) block_dim=adaptive {
        <i32 x1> $10 = local load [ [$5[0]]]
        <i32 x1> $11 = local load [ [$3[0]]]
        <i32 x1> $12 = add $11 $10
        <i32 x1> $13 : local store [$3 <- $10]
        <i32 x1> $14 : local store [$5 <- $12]
      }
    }
    <i32 x1> $15 = local load [ [$5[0]]]
    <i32 x1> $16 : kernel return $15
  }
}
```

**Comment by archibate** [comment]
> At least on Metal, this broke

Is this because some configs not sync with LLVM ones on Metal?
Sorry... could not verify OpenGL behavior on my ancient laptop, @yuanming-hu Could you do me a favor?

I think ultimately we should:

1. Have an option to explicitly disable range-for from being parallelized instead of ad-hoc `if 1:`, e.g.:

```py
@ti.kernel
def func():
  for i in ti.serial(range(4)):
    print(i)
```

2. Have an possibility to make non-top-level range-for being parallelized, which could be useful in some situations like:

vertex shader (parallelize among vertices&faces) -> fragment shader (parallelize among pixels in each face)

**Comment by k-ye** [comment]
> Is this because some configs not sync with LLVM ones on Metal?

I don't think so, because the IR itself already seems buggy. (I don't think CI has test on CUDA, does it?)

> Have an option to explicitly disable range-for from being parallelized instead of ad-hoc if 1:, e.g.:

+1. This will be a lot less error-prone.

> Have an possibility to make non-top-level range-for being parallelized, which could be useful in some situations like:

I feel like this is a very non-trivial work. It fundamentally changes how Taichi produces the backend code. In certain cases, maybe Taichi can be made smart enough to figure out how to flatten out a nested loop into the top level. Still, that will likely involve some more IR analysis and transformation works.

**Pull request merged**